clearcode/store_scans.py

```diff
@@ -152,38 +161,76 @@ def get_repo_name(purl: PackageURL, length: int=3) -> str:
-    then take the latest pull. Return the repo object"""
+def get_or_init_repo(repo_name: str, work_dir: Path, repo_namespace: str= "", pull=False):
+    """
+    Return a repo object for repo name and namespace 
+    and store it in the work dir. Clone if it does not 
+    exist optionally take the latest pull if it does exist.
+    """
+    # TODO: Manage org repo name
+    # MAYBE: CREATE ALL THE REPOS AT A TIME AND CLONE THEM LOCALLY
+    if repo_name not in get_github_repos(user_name="TG1999"):
+        repo_url = create_github_repo(repo_name=repo_name)
     repo_path = work_dir / repo_name
     if repo_path.exists():
-        return Repo(repo_path)
+        repo = Repo(repo_path)
+        if pull:
+            repo.origin.pull()
     else:
-        repo = Repo.init(repo_path)
-        return repo
+        repo = Repo.clone_from(repo_url, repo_path)
+    return repo
 
 
-def get_scan_download_url(purl:str, github_org:str, scan_file_name: str = "scancode-toolkit-scan.json"):
-    repo_name = get_repo_name(purl=purl)
+def get_scan_download_url(namespace:str, purl:str, scan_file_name: str = "scancode-toolkit-scan.json"):
+    purl_hash = get_purl_hash(purl=purl)
     purl_path = get_purl_path(purl)
-    return f"https://raw.githubusercontent.com/{github_org}/{repo_name}/main/{purl_path}/{scan_file_name}"
+    return f"https://raw.githubusercontent.com/{namespace}/{purl_hash}/main/{purl_path}/{scan_file_name}"
+
+
+def create_github_repo(repo_name, token=os.getenv("GH_TOKEN")):
+    headers = {
+        'Authorization': f'token {token}',
+        'Accept': 'application/vnd.github.v3+json'
+    }
+
+    data = {
+        'name': repo_name,
+    }
+
+    url = 'https://api.github.com/user/repos'
+
+    response = requests.post(url, headers=headers, json=data)
 
+    if response.status_code == 201:
+        print(f"Repository '{repo_name}' created successfully!")
+    else:
+        print(f"Failed to create repository. Status code: {response.status_code}")
+        print(response.text)
+
+
+def get_github_repos(user_name, token=os.getenv("GH_TOKEN")):
+    """
+    Yield full repo names for a user or org name, use the optional ``token`` if provided.
+    Full repo name is in the form user or org name / repo name
+    """
+    headers = {
+        'Accept': 'application/vnd.github.v3+json'
+    }
+    if token:
+        headers['Authorization'] = f'token {token}'
+
+    url = f'https://api.github.com/users/{user_name}/repos'
+    response = requests.get(url, headers=headers)
+
+    # TODO: We need have a way to handle failures from GH API
+    if not response.status_code == 200:
+        raise Exception(f"HTTP {response.status_code}: Failed to get repos for {user_name}")
+
+    data = response.json()
+    for repo_data in data:
+        full_repo_name = repo_data.get("full_name")
+        if full_repo_name:
+            yield full_repo_name
 
 if __name__ == "__main__":
```

Comment:
Create a management command out of this, where we can specify the directory, github org, and count as options